packages/ciphernode/net/src/dialer.rs

```diff
@@ -1,210 +1,261 @@
-pub async fn dial_peers(
-    cmd_tx: &mpsc::Sender<NetworkPeerCommand>,
-    event_tx: &broadcast::Sender<NetworkPeerEvent>,
-    peers: &Vec<String>,
-) -> Result<()> {
-    let futures: Vec<_> = peers
-        .iter()
-        .map(|addr| dial_multiaddr(cmd_tx, event_tx, addr))
-        .collect();
-    let results = join_all(futures).await;
-    results.into_iter().for_each(trace_error);
-    Ok(())
+#[derive(Clone)]
+struct PendingConnection {
+    addr: String,
+    attempt: u32,
+    delay_ms: u64,
 }
 
-/// Attempt a connection with retrys to a multiaddr return an error if the connection could not be resolved after the retries.
-async fn attempt_connection(
-    cmd_tx: &mpsc::Sender<NetworkPeerCommand>,
-    event_tx: &broadcast::Sender<NetworkPeerEvent>,
-    multiaddr: &Multiaddr,
-) -> Result<(), RetryError> {
-    let mut event_rx = event_tx.subscribe();
-    let multi = get_resolved_multiaddr(multiaddr).map_err(to_retry)?;
-    let opts: DialOpts = multi.clone().into();
-    let dial_connection = opts.connection_id();
-    info!("Dialing: '{}' with connection '{}'", multi, dial_connection);
-    cmd_tx
-        .send(NetworkPeerCommand::Dial(opts))
-        .await
-        .map_err(to_retry)?;
-    wait_for_connection(&mut event_rx, dial_connection).await
+#[derive(Clone)]
+pub struct DialerActor {
+    net_bus: Addr<EventBus<NetworkPeerEvent>>,
+    tx: mpsc::Sender<NetworkPeerCommand>,
+    pending_connection: HashMap<ConnectionId, PendingConnection>,
 }
 
-/// Wait for results of a retry based on a given correlation id and return the correct variant of
-/// RetryError depending on the result from the downstream event
-async fn wait_for_connection(
-    event_rx: &mut broadcast::Receiver<NetworkPeerEvent>,
-    dial_connection: ConnectionId,
-) -> Result<(), RetryError> {
-    loop {
-        // Create a timeout future that can be reset
-        select! {
-            result = event_rx.recv() => {
-                match result.map_err(to_retry)? {
-                    NetworkPeerEvent::ConnectionEstablished { connection_id } => {
-                        if connection_id == dial_connection {
-                            info!("Connection Established");
-                            return Ok(());
-                        }
+impl DialerActor {
+    pub fn new(
+        net_bus: Addr<EventBus<NetworkPeerEvent>>,
+        tx: mpsc::Sender<NetworkPeerCommand>,
+    ) -> Addr<Self> {
+        let addr = Self {
+            net_bus: net_bus.clone(),
+            tx,
+            pending_connection: HashMap::new(),
+        }
+        .start();
+
+        // Listen on all events
+        net_bus.do_send(Subscribe {
+            event_type: String::from("*"),
+            listener: addr.clone().recipient(),
+        });
+
+        addr
+    }
+
+    pub fn dial_peer(
+        addr: String,
+        net_bus: Addr<EventBus<NetworkPeerEvent>>,
+        tx: mpsc::Sender<NetworkPeerCommand>,
```

Comment:
Why do we need a `tx` here? Why can we not simply use Actor messages? We probably need: 

```
enum NetworkMessage {
  Event(NetworkPeerEvent),
  Command(NetworkPeerCommand)
}
```

or something similar.

Reply:
They don't play well as the trait `Clone` is not implemented for `DialOpts` inside `NetworkPeerCommand`.
Which is required by the `Event` Trait and is not removeable. 
So the easiest solution was to use a mpsc tx

Reply:
Can we avoid sending DialOpts in an event and instead send something serializable?

Reply:
Might be able to make it send `Arc<DialOpts>` instead?

Reply:
Also can use `CorrelationId` instead of `ConnectionId` from `DialOpts`

Reply:
Yeah sending `Arc<DialOpts>` works

Reply:
We could do that by moving the address resolution logic to the network peer instead of the dialer. 
I guess thats a small cost to pay for a whole lot of simplification.

Reply:
> Yeah sending `Arc<DialOpts>` works

Can't clone it in network peer `self.swarm.dial((*dial_opts).clone())`